storage/flush.go

```diff
@@ -35,63 +34,94 @@ import (
 )
 
 var (
-	errFlushAlreadyInProgress = errors.New("flush already in progress")
+	errFlushOrSnapshotAlreadyInProgress = errors.New("flush or snapshot already in progress")
 )
 
 type flushManager struct {
 	sync.RWMutex
 
-	database        database
-	opts            Options
-	nowFn           clock.NowFn
-	pm              persist.Manager
-	flushInProgress bool
-	status          tally.Gauge
+	database database
+	opts     Options
+	pm       persist.Manager
+	// isFlushingOrSnapshotting is used to protect the flush manager against
+	// concurrent use, while flushInProgress and snapshotInProgress are more
+	// granular and are used for emitting granular gauges.
+	isFlushingOrSnapshotting bool
+	flushInProgress          bool
+	snapshotInProgress       bool
+	isFlushing               tally.Gauge
+	isSnapshotting           tally.Gauge
 }
 
 func newFlushManager(database database, scope tally.Scope) databaseFlushManager {
 	opts := database.Options()
 	return &flushManager{
-		database: database,
-		opts:     opts,
-		nowFn:    opts.ClockOptions().NowFn(),
-		pm:       opts.PersistManager(),
-		status:   scope.Gauge("flush"),
+		database:       database,
+		opts:           opts,
+		pm:             opts.PersistManager(),
+		isFlushing:     scope.Gauge("flush"),
+		isSnapshotting: scope.Gauge("snapshot"),
 	}
 }
 
 func (m *flushManager) Flush(curr time.Time) error {
 	// ensure only a single flush is happening at a time
 	m.Lock()
-	if m.flushInProgress {
+	if m.isFlushingOrSnapshotting {
 		m.Unlock()
-		return errFlushAlreadyInProgress
+		return errFlushOrSnapshotAlreadyInProgress
 	}
-	m.flushInProgress = true
+	m.isFlushingOrSnapshotting = true
 	m.Unlock()
 
+	defer func() {
+		m.Lock()
+		m.isFlushingOrSnapshotting = false
+		m.Unlock()
+	}()
+
 	// create flush-er
 	flush, err := m.pm.StartFlush()
 	if err != nil {
 		return err
 	}
 
-	defer func() {
-		m.Lock()
-		m.flushInProgress = false
-		m.Unlock()
-	}()
-
 	namespaces, err := m.database.GetOwnedNamespaces()
 	if err != nil {
 		return err
 	}
 
 	multiErr := xerrors.NewMultiError()
+	m.setFlushInProgress(true)
 	for _, ns := range namespaces {
+		// Flush first because we will only snapshot if there are no outstanding flushes
 		flushTimes := m.namespaceFlushTimes(ns, curr)
 		multiErr = multiErr.Add(m.flushNamespaceWithTimes(ns, flushTimes, flush))
 	}
+	m.setFlushInProgress(false)
+
+	// Perform two separate loops through all the namespaces so that we can emit better
```

Comment:
hm, is it safe to re-use the `flush` object across flush/snapshot once it's been used?

Reply:
Yeah I think so, if you look at the StartFlush method all it does is protect against concurrent flushes